src/sys/windows/afd.rs

```diff
@@ -0,0 +1,222 @@
+                    RtlNtStatusToDosError(status) as i32
+                ));
+            }
+            let fd = File::from_raw_handle(afd_helper_handle as RawHandle);
+            let afd = Afd { fd };
+            let token = NEXT_TOKEN.fetch_add(1, Ordering::Relaxed) + 1;
+            cp.add_handle(token, &afd.fd)?;
+            match SetFileCompletionNotificationModes(
+                afd_helper_handle,
+                FILE_SKIP_SET_EVENT_ON_HANDLE,
+            ) {
+                0 => Err(io::Error::last_os_error()),
+                _ => Ok(afd),
+            }
+        }
+    }
+
+    /// Poll `Afd` instance with `AfdPollInfo`.
+    ///
+    /// # Unsafety
+    ///
+    /// This function is unsafe due to memory of `IO_STATUS_BLOCK` still being used by `Afd` instance while `Ok(false)` (`STATUS_PENDING`).
+    /// `iosb` needs to be untouched after the call while operation is in effective at ALL TIME except for `cancel` method.
+    /// So be careful not to `poll` twice while polling.
+    pub unsafe fn poll(
+        &self,
+        info: &mut AfdPollInfo,
+        iosb: Pin<&mut IO_STATUS_BLOCK>,
+        overlapped: PVOID,
+    ) -> io::Result<bool> {
+        let info_ptr: PVOID = info as *mut _ as PVOID;
+        let iosb_ptr = iosb.get_mut();
+        (*iosb_ptr).u.Status = STATUS_PENDING;
+        let status = NtDeviceIoControlFile(
+            self.fd.as_raw_handle(),
+            null_mut(),
+            None,
+            overlapped,
+            iosb_ptr,
+            IOCTL_AFD_POLL,
+            info_ptr,
+            size_of::<AfdPollInfo>() as u32,
+            info_ptr,
+            size_of::<AfdPollInfo>() as u32,
+        );
+        match status {
+            STATUS_SUCCESS => Ok(true),
+            STATUS_PENDING => Ok(false),
+            _ => Err(io::Error::from_raw_os_error(
+                RtlNtStatusToDosError(status) as i32
+            )),
+        }
+    }
+
+    /// Cancel previous polled request of `Afd`.
+    ///
+    /// iosb needs to be used by `poll` first for valid `cancel`.
+    ///
+    /// # Unsafety
+    ///
+    /// This function is unsafe due to memory of `IO_STATUS_BLOCK` still being used by `Afd` instance while `Ok(false)` (`STATUS_PENDING`).
+    /// Use it only with request is still being polled so that you have valid `IO_STATUS_BLOCK` to use.
+    pub unsafe fn cancel(&self, iosb: Pin<&mut IO_STATUS_BLOCK>) -> io::Result<()> {
+        if iosb.u.Status != STATUS_PENDING {
+            return Ok(());
+        }
+
+        let mut cancel_iosb: IO_STATUS_BLOCK = std::mem::zeroed();
+        let status = NtCancelIoFileEx(self.fd.as_raw_handle(), iosb.get_mut(), &mut cancel_iosb);
+        if status == STATUS_SUCCESS || status == STATUS_NOT_FOUND {
+            return Ok(());
+        }
+        Err(io::Error::from_raw_os_error(
+            RtlNtStatusToDosError(status) as i32
+        ))
+    }
+}
+
+pub const AFD_POLL_RECEIVE: u32 = 0x0001;
```

Comment:
Don't any Windows API crate export these? If fine with defining them here, but I would rather some other crate that care of them, e.g. `winapi`.

Reply:
This is internal component in AFD which is not public API of Windows.

Reply:
I think this cannot be in winapi.

Reply:
@Thomasdezeeuw Yeah, AFD isn't **technically** a public API, so the `winapi` crate doesn't include it.